src/transformStateWithClones.js

```diff
@@ -6,8 +6,43 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  for (const action of actions) {
+    const { type = null, extraData = {}, keysToRemove = [] } = action;
+
+    let newObj;
+
+    if (states.length === 0) {
+      newObj = { ...state };
+    } else {
+      newObj = { ...states[states.length - 1] };
+    }
+
+    switch (type) {
+      case 'addProperties':
+        newObj = Object.assign(newObj, extraData);
+        break;
+      case 'removeProperties':
+        keysToRemove.forEach((element) => {
+          if (newObj[element]) {
+            delete newObj[element];
```

Comment:
The check `if (newObj[element])` is not needed. You can delete a property from an object regardless of its existence. If the property does not exist, the delete operation will have no effect.